generators/beagle/src/main/scala/BeaglePeriphery.scala

```diff
@@ -22,85 +22,157 @@ case object PeripheryBeagleKey extends Field[BeagleParams]
+
+  val hbwif_rsts   = Output(Vec(p(HbwifNumLanes), Bool()))
+  val switcher_sel = Output(Bool())
 }
 
-trait HasBeagleTopModuleContents extends MultiIOModule with HasRegMap {
+/**
+ * MMIO controlled components
+ */
+trait HasBeagleTopModuleContents extends MultiIOModule with HasRegMap
+{
   val io: HasBeagleTopBundleContents
   implicit val p: Parameters
   def params: BeagleParams
   def c = params
 
-  val clk_sel = withReset(io.rst_async) {
-    Module(new AsyncResetRegVec(w = c.clkSelBits, init = 0))
+  // main clock divider regs
+  val r_uncore_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.uncoreClkDivBits, init = c.uncoreClkDivInit))
+  }
+
+  val r_bh_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.bhClkDivBits, init = c.bhClkDivInit))
+  }
+
+  val r_rs_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.rsClkDivBits, init = c.rsClkDivInit))
+  }
+
+  // output clock divider regs
+  val r_bh_clk_out_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.bhClkDivBits, init = c.bhClkDivInit))
+  }
+
+  val r_rs_clk_out_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.rsClkDivBits, init = c.rsClkDivInit))
+  }
+
+  // lbwif clock divider reg
+  val r_lbwif_clk_divisor = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = c.lbwifClkDivBits, init = c.lbwifClkDivInit))
   }
-  val lbwif_divider = withReset(io.rst_async) {
-    Module(new AsyncResetRegVec(w = c.lbwifDividerBits, init = p(LbwifDividerInit)))
+
+  // uncore select bit reg
+  val r_uncore_clk_pass_sel = withReset(io.rst_async) {
+    Module(new AsyncResetRegVec(w = 1, init = 0))
   }
-  val hbwif_rsts = Seq.fill(p(HbwifNumLanes)) {
+
+  // hbwif reset signals
+  val r_hbwif_rsts = Seq.fill(p(HbwifNumLanes)) {
     withReset(io.rst_async) {
       Module(new AsyncResetRegVec(w = 1, init = 1))
     }
   }
-  val switcher_sel = withReset(io.rst_async) {
+
+  // connect hbwif or lbwif to memory system
+  val r_switcher_sel = withReset(io.rst_async) {
     Module(new AsyncResetRegVec(w = 1, init = 1))
   }
 
-  io.clk_sel := clk_sel.io.q
-  io.lbwif_divider := lbwif_divider.io.q
-  io.hbwif_rsts := VecInit(hbwif_rsts.map(_.io.q))
-  io.switcher_sel := switcher_sel.io.q
+  // connect to io out of scr
+  io.uncore_clk_divisor  := r_uncore_clk_divisor.io.q
+  io.bh_clk_divisor      := r_bh_clk_divisor.io.q
+  io.rs_clk_divisor      := r_rs_clk_divisor.io.q
+  io.bh_clk_out_divisor  := r_bh_clk_out_divisor.io.q
+  io.rs_clk_out_divisor  := r_rs_clk_out_divisor.io.q
+  io.lbwif_clk_divisor   := r_lbwif_clk_divisor.io.q
+  io.uncore_clk_pass_sel := r_uncore_clk_pass_sel.io.q
+  io.hbwif_rsts          := VecInit(r_hbwif_rsts.map(_.io.q))
+  io.switcher_sel        := r_switcher_sel.io.q
+
+  // connect to mmio
+
+  // needed to space out the mmio correctly
+  require(c.uncoreClkDivBits == 8)
```

Comment:
This doesn't seem necessary. You can just have the reg map space them out right?

Reply:
It serves as a reminder that you need to space it out correctly. But no, this is not strictly necessary.